### PR TITLE
feat: LT-1980 MOD Clearances

### PR DESCRIPTION
**Pull request by BenAndrew-Kainos**
**Adding new MOD Application types: F680 & Gifting**

Key API changes:

- Add new application types, unit tests & serializers
- Tidy up creators methods for validating applications
- Update MOD licence generator in test client
- Resolved issue with party error messages

### CODE REVIEW

File: applications/libraries/get_applications.py [hunk 1]
```diff
@@ -26,6 +28,10 @@ def get_application(pk, organisation_id=None):
             return HmrcQuery.objects.get(pk=pk)
         elif application_type == ApplicationType.EXHIBITION_CLEARANCE:
             return ExhibitionClearanceApplication.objects.get(pk=pk)
+        elif application_type == ApplicationType.GIFTING_CLEARANCE:
+            return GiftingClearanceApplication.objects.get(pk=pk)
+        elif application_type == ApplicationType.F_680_CLEARANCE:
```
Comment (jfaracik) [inline applications/libraries/get_applications.py]: Shouldn't that be `F680_CLEARANCE`?

Comment (BenAndrew-Kainos) [reply applications/libraries/get_applications.py]: Excellent point

File: applications/tests/tests_mod_clearance_submit.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.json()["application"]["name"], self.draft.name)
+
+    def test_submit_F680_with_end_user_and_without_third_party_success(self):
+        self.draft.third_parties.all().delete()
+
+        response = self.client.put(self.url, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.json()["application"]["name"], self.draft.name)
```
Comment (efahy) [inline applications/tests/tests_mod_clearance_submit.py]: Perhaps for this test and the next, assert that the party of the correct type exists?

File: applications/tests/tests_mod_clearance_submit.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    def test_submit_MOD_clearance_without_goods_failure(self):
+        GoodOnApplication.objects.get(application=self.draft).delete()
+
+        response = self.client.put(self.url, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.json()["errors"]["goods"], strings.Applications.Standard.NO_GOODS_SET)
+
+    def test_submit_MOD_clearance_without_parties_failure(self):
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/tests/tests_mod_clearance_submit.py]: Yeah you're right. Will remove

File: applications/tests/tests_mod_clearance_submit.py (hunk 2 above)
Comment (BenAndrew-Kainos) [reply applications/tests/tests_mod_clearance_submit.py]: Do you mean doesn't exist?

File: applications/tests/tests_create_application.py [hunk 4]
```diff
@@ -55,6 +57,38 @@ def test_create_draft_exhibition_clearance_application_successful(self):
...
+
+        data = {
+            "name": "Test",
+            "application_type": ApplicationType.GIFTING_CLEARANCE,
+        }
+
+        response = self.client.post(self.url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(GiftingClearanceApplication.objects.count(), 1)
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/tests/tests_create_application.py]: done

File: applications/serializers/f680_clearance.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            "application_type",
+            "organisation",
+            "type",
+            "status",
+        )
+
+
+class F680ClearanceUpdateSerializer(GenericApplicationUpdateSerializer):
+    name = CharField(
+        max_length=100,
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/serializers/f680_clearance.py]: I suggested this before and everyone told me no!

[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/serializers/f680_clearance.py]: I'll raise tommorow

File: applications/creators.py [hunk 6]
```diff
@@ -112,11 +102,62 @@ def _validate_standard_licence(draft, errors):
...
 
+def _validate_gifting_clearance(draft, errors):
+    errors = _validate_end_user(draft, errors)
+    errors = _validate_third_parties(draft, errors)
+    errors = _validate_has_goods(draft, errors)
+
+    return errors
+
+
+def _validate_f680_clearance(draft, errors):
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/creators.py]: I'll implement this

File: applications/tests/tests_mod_clearance_submit.py (hunk 2 above)
Comment (efahy) [reply applications/tests/tests_mod_clearance_submit.py]: No I meant exists, i.e. in this case there's an end user on the application.  But if you think this is overkill I won't argue

File: applications/creators.py [hunk 7]
```diff
@@ -112,11 +102,62 @@ def _validate_standard_licence(draft, errors):
     return errors
 
 
+def _validate_third_parties(draft, errors):
```
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/creators.py]: Added docstrings

File: applications/serializers/f680_clearance.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/serializers/f680_clearance.py]: Doing now

[the remark this answers is not in the document]
Comment (BenAndrew-Kainos) [reply applications/serializers/f680_clearance.py]: As discussed people aren't happy with it so not at the moment

File: applications/tests/tests_mod_clearance_submit.py (hunk 2 above)
Comment (BenAndrew-Kainos) [reply applications/tests/tests_mod_clearance_submit.py]: Done

File: parties/serializers.py [hunk 8]
```diff
@@ -12,14 +12,14 @@
 
 
 class PartySerializer(serializers.ModelSerializer):
-    name = serializers.CharField()
-    address = serializers.CharField()
-    country = CountrySerializerField()
+    name = serializers.CharField(error_messages={"required": Parties.NULL_NAME})
```
Comment (jfaracik) [inline parties/serializers.py]: Good change 👍

Comment (BenAndrew-Kainos) [reply parties/serializers.py]: Thanks

**Pull request merged**